tfhe/src/core_crypto/entities/cleartext.rs

```diff
@@ -1,7 +1,79 @@
 //! Module containing the definition of the Cleartext.
 
-use crate::core_crypto::commons::numeric::Numeric;
+use crate::core_crypto::commons::traits::*;
 
 /// A cleartext, not encoded, value.
-#[derive(Clone, Copy, Debug, serde::Serialize, serde::Deserialize)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
 pub struct Cleartext<T: Numeric>(pub T);
+/// An immutable reference to a cleartext value.
+///
+/// Can be converted to a cleartext via a call to `into`
+/// ```
+/// use tfhe::core_crypto::entities::*;
+///
+/// pub fn takes_cleartext(clear: Cleartext<u64>) {
+///     println!("{clear:?}");
+/// }
+///
+/// let encoded_msg = 3u64 << 60;
+///
+/// // A cleartext containing a reference can be returned by iterators for example, here is how
+/// // to convert them painlessly.
+/// let ref_cleartext = CleartextRef(&encoded_msg);
+/// takes_cleartext(ref_cleartext.into());
+/// ```
+pub struct CleartextRef<'data, T: Numeric>(pub &'data T);
+/// A mutable reference to a cleartext (encoded) value.
+///
+/// Can be converted to a cleartext via a call to `into`
+/// ```
+/// use tfhe::core_crypto::entities::*;
+///
+/// pub fn takes_cleartext(clear: Cleartext<u64>) {
+///     println!("{clear:?}");
+/// }
+///
+/// let mut encoded_msg = 3u64 << 60;
+///
+/// // A cleartext containing a reference can be returned by iterators for example, here is how
+/// // to convert them painlessly.
+/// let ref_cleartext = CleartextRefMut(&mut encoded_msg);
+/// takes_cleartext(ref_cleartext.into());
+/// ```
+pub struct CleartextRefMut<'data, T: Numeric>(pub &'data mut T);
+
+impl<'data, T: Numeric> CreateFrom<&'data [T]> for CleartextRef<'data, T> {
+    type Metadata = ();
+
+    #[inline]
+    fn create_from(from: &[T], _: Self::Metadata) -> CleartextRef<'_, T> {
+        CleartextRef(&from[0])
+    }
+}
+
+impl<'data, T: Numeric> CreateFrom<&'data mut [T]> for CleartextRefMut<'data, T> {
+    type Metadata = ();
+
+    #[inline]
+    fn create_from(from: &mut [T], _: Self::Metadata) -> CleartextRefMut<'_, T> {
+        CleartextRefMut(&mut from[0])
+    }
+}
+
+impl<T: Numeric> From<CleartextRef<'_, T>> for Cleartext<T>
+where
+    T: Copy,
+{
+    fn from(cleartext_ref: CleartextRef<T>) -> Cleartext<T> {
+        Cleartext(*cleartext_ref.0)
+    }
+}
+
+impl<T: Numeric> From<CleartextRefMut<'_, T>> for Cleartext<T>
+where
+    T: Copy,
+{
+    fn from(cleartext_ref: CleartextRefMut<T>) -> Cleartext<T> {
+        Cleartext(*cleartext_ref.0)
+    }
```

Comment:
Technically, I think implementing ToOwned would be a better fit
https://doc.rust-lang.org/std/borrow/trait.ToOwned.html#required-methods

Reply:
good point, it's just the pattern we had for Plaintext (that I also did 🙄 )

Reply:
ah damn forgot to update this -_-

Reply:
is it fine to have both ?

Reply:
A yes Borrow requires to return a ref so can't really use it